File: implementations/cpp/ropes_gap_buffers_piece_tables.cpp
/**
 * @file ropes_gap_buffers_piece_tables.cpp
 * @brief Reference implementations of Gap Buffers and Piece Tables for Text Editing.
 *
 * Implements:
 * 1. GapBuffer: Localized typing with moving gap and dynamic resizing.
 * 2. PieceTable: Dual-buffer (original/add) piece table with piece coalescing.
 * 3. Differential testing against std::string reference oracle.
 * Verified with C++17 (-std=c++17 -O3 -Wall -Wextra -Werror).
 */

#include <iostream>
#include <vector>
#include <string>
#include <cassert>
#include <algorithm>
#include <random>

namespace dsa {

// ============================================================================
// 1. Gap Buffer
// ============================================================================

class GapBuffer {
private:
    std::vector<char> buffer_;
    size_t gap_start_ = 0;
    size_t gap_end_ = 16;
    size_t capacity_ = 16;

    void grow(size_t min_needed) {
        size_t new_cap = std::max(capacity_ * 2, capacity_ + min_needed + 16);
        std::vector<char> new_buf(new_cap);

        // Copy prefix before gap
        for (size_t i = 0; i < gap_start_; ++i) {
            new_buf[i] = buffer_[i];
        }

        // Copy suffix after gap to the end of new buffer
        size_t suffix_len = capacity_ - gap_end_;
        size_t new_gap_end = new_cap - suffix_len;
        for (size_t i = 0; i < suffix_len; ++i) {
            new_buf[new_gap_end + i] = buffer_[gap_end_ + i];
        }

        buffer_ = std::move(new_buf);
        gap_end_ = new_gap_end;
        capacity_ = new_cap;
    }

public:
    explicit GapBuffer(size_t initial_cap = 16)
        : buffer_(initial_cap), gap_start_(0), gap_end_(initial_cap), capacity_(initial_cap) {}

    size_t size() const {
        return capacity_ - (gap_end_ - gap_start_);
    }

    size_t cursor() const {
        return gap_start_;
    }

    void move_cursor(size_t pos) {
        assert(pos <= size());
        if (pos < gap_start_) {
            // Shift gap left
            size_t shift = gap_start_ - pos;
            while (shift > 0) {
                gap_start_--;
                gap_end_--;
                buffer_[gap_end_] = buffer_[gap_start_];
                shift--;
            }
        } else if (pos > gap_start_) {
            // Shift gap right
            size_t shift = pos - gap_start_;
            while (shift > 0) {
                buffer_[gap_start_] = buffer_[gap_end_];
                gap_start_++;
                gap_end_++;
                shift--;
            }
        }
    }

    void insert(char c) {
        if (gap_start_ == gap_end_) {
            grow(1);
        }
        buffer_[gap_start_++] = c;
    }

    void insert(const std::string& str) {
        if (gap_end_ - gap_start_ < str.size()) {
            grow(str.size());
        }
        for (char c : str) {
            buffer_[gap_start_++] = c;
        }
    }

    // Backspace: delete character before cursor
    bool erase_back() {
        if (gap_start_ == 0) return false;
        gap_start_--;
        return true;
    }

    std::string to_string() const {
        std::string res;
        res.reserve(size());
        for (size_t i = 0; i < gap_start_; ++i) {
            res.push_back(buffer_[i]);
        }
        for (size_t i = gap_end_; i < capacity_; ++i) {
            res.push_back(buffer_[i]);
        }
        return res;
    }
};

// ============================================================================
// 2. Piece Table
// ============================================================================

enum class BufferType { ORIGINAL, ADD };

struct Piece {
    BufferType buffer;
    size_t start;
    size_t length;
};

class PieceTable {
private:
    std::string original_;
    std::string add_;
    std::vector<Piece> pieces_;

public:
    explicit PieceTable(std::string initial_text = "")
        : original_(std::move(initial_text)) {
        if (!original_.empty()) {
            pieces_.push_back({BufferType::ORIGINAL, 0, original_.size()});
        }
    }

    size_t size() const {
        size_t total = 0;
        for (const auto& p : pieces_) total += p.length;
        return total;
    }

    void insert(size_t pos, const std::string& text) {
        if (text.empty()) return;
        assert(pos <= size());

        size_t add_offset = add_.size();
        add_.append(text);
        size_t text_len = text.size();

        if (pieces_.empty()) {
            pieces_.push_back({BufferType::ADD, add_offset, text_len});
            return;
        }

        // Fast coalescing optimization: sequential typing at end
        if (pos == size() && pieces_.back().buffer == BufferType::ADD &&
            pieces_.back().start + pieces_.back().length == add_offset) {
            pieces_.back().length += text_len;
            return;
        }

        size_t curr_pos = 0;
        for (size_t i = 0; i < pieces_.size(); ++i) {
            size_t next_pos = curr_pos + pieces_[i].length;

            if (pos == curr_pos) {
                // Insert directly before piece i
                pieces_.insert(pieces_.begin() + i, {BufferType::ADD, add_offset, text_len});
                return;
            } else if (pos > curr_pos && pos < next_pos) {
                // Split piece i into two and insert in between
                size_t left_len = pos - curr_pos;
                size_t right_len = pieces_[i].length - left_len;

                Piece left_piece = {pieces_[i].buffer, pieces_[i].start, left_len};
                Piece new_piece = {BufferType::ADD, add_offset, text_len};
                Piece right_piece = {pieces_[i].buffer, pieces_[i].start + left_len, right_len};

                pieces_[i] = left_piece;
                pieces_.insert(pieces_.begin() + i + 1, {new_piece, right_piece});
                return;
            }
            curr_pos = next_pos;
        }

        // Insert at the very end
        pieces_.push_back({BufferType::ADD, add_offset, text_len});
    }

    void erase(size_t pos, size_t length) {
        if (length == 0 || pieces_.empty()) return;
        assert(pos + length <= size());

        size_t curr_pos = 0;
        std::vector<Piece> new_pieces;

        for (const auto& p : pieces_) {
            size_t p_start = curr_pos;
            size_t p_end = curr_pos + p.length;

            if (p_end <= pos || p_start >= pos + length) {
                // Completely outside deleted region: keep unchanged
                new_pieces.push_back(p);
            } else {
                // Partially or completely inside deleted region
                if (pos > p_start) {
                    // Left intact portion
                    new_pieces.push_back({p.buffer, p.start, pos - p_start});
                }
                if (pos + length < p_end) {
                    // Right intact portion
                    size_t skipped = (pos + length) - p_start;
                    new_pieces.push_back({p.buffer, p.start + skipped, p.length - skipped});
                }
            }
            curr_pos = p_end;
        }

        pieces_ = std::move(new_pieces);
    }

    std::string to_string() const {
        std::string res;
        res.reserve(size());
        for (const auto& p : pieces_) {
            const std::string& buf = (p.buffer == BufferType::ORIGINAL) ? original_ : add_;
            res.append(buf, p.start, p.length);
        }
        return res;
    }
};

} // namespace dsa

int main() {
    std::cout << "Running Ropes, Gap Buffers, and Piece Tables verification..." << std::endl;

    // 1. Gap Buffer Basic Test
    dsa::GapBuffer gb(8);
    gb.insert("Hello");
    gb.insert(' ');
    gb.insert("World");
    assert(gb.to_string() == "Hello World");

    gb.move_cursor(5); // Cursor right after "Hello"
    gb.insert(" Beautiful");
    assert(gb.to_string() == "Hello Beautiful World");

    gb.move_cursor(gb.size());
    gb.erase_back(); // Delete 'd'
    assert(gb.to_string() == "Hello Beautiful Worl");

    // 2. Piece Table Basic Test
    dsa::PieceTable pt("The brown fox");
    pt.insert(4, "quick ");
    assert(pt.to_string() == "The quick brown fox");

    pt.insert(pt.size(), " jumps");
    assert(pt.to_string() == "The quick brown fox jumps");

    pt.erase(4, 6); // Delete "quick "
    assert(pt.to_string() == "The brown fox jumps");

    // 3. Differential Fuzzing vs std::string Oracle
    std::string oracle = "Initial Document Content for Testing";
    dsa::PieceTable fuzz_pt(oracle);
    dsa::GapBuffer fuzz_gb(64);
    fuzz_gb.insert(oracle);

    std::mt19937 rng(42);
    const int OPS = 500;

    for (int step = 0; step < OPS; ++step) {
        int op = rng() % 2;

        if (op == 0) {
            // Insert random snippet
            size_t pos = oracle.empty() ? 0 : rng() % (oracle.size() + 1);
            std::string text = " [edit_" + std::to_string(step) + "] ";

            oracle.insert(pos, text);
            fuzz_pt.insert(pos, text);

            fuzz_gb.move_cursor(pos);
            fuzz_gb.insert(text);
        } else {
            // Erase range
            if (!oracle.empty()) {
                size_t pos = rng() % oracle.size();
                size_t max_len = std::min<size_t>(oracle.size() - pos, 8);
                size_t len = (rng() % max_len) + 1;

                oracle.erase(pos, len);
                fuzz_pt.erase(pos, len);

                fuzz_gb.move_cursor(pos + len);
                for (size_t i = 0; i < len; ++i) {
                    fuzz_gb.erase_back();
                }
            }
        }

        assert(fuzz_pt.size() == oracle.size());
        assert(fuzz_gb.size() == oracle.size());
        assert(fuzz_pt.to_string() == oracle);
        assert(fuzz_gb.to_string() == oracle);
    }

    std::cout << "[PASS] Basic gap buffer typing, cursor movement, and deletion verified." << std::endl;
    std::cout << "[PASS] Basic piece table multi-piece insertion and deletion verified." << std::endl;
    std::cout << "[PASS] 500 differential fuzzing operations matched std::string bit-for-bit." << std::endl;
    std::cout << "All Ropes, Gap Buffers, and Piece Tables assertions passed successfully!" << std::endl;
    return 0;
}

File: implementations/python/ropes_gap_buffers_piece_tables.py
"""
Reference implementations of Gap Buffers and Piece Tables for Text Editing.

Implements:
1. GapBuffer: Localized typing with moving gap and dynamic resizing.
2. PieceTable: Dual-buffer (original/add) piece table with piece coalescing.
3. Differential testing against Python string oracle.
"""

import random
import unittest
from typing import List


class GapBuffer:
    """Gap Buffer for localized cursor text editing."""

    def __init__(self, initial_capacity: int = 16):
        self.capacity = initial_capacity
        self.buffer = [''] * self.capacity
        self.gap_start = 0
        self.gap_end = self.capacity

    def __len__(self) -> int:
        return self.capacity - (self.gap_end - self.gap_start)

    def _grow(self, min_needed: int) -> None:
        new_cap = max(self.capacity * 2, self.capacity + min_needed + 16)
        new_buf = [''] * new_cap

        # Copy prefix
        for i in range(self.gap_start):
            new_buf[i] = self.buffer[i]

        # Copy suffix to end
        suffix_len = self.capacity - self.gap_end
        new_gap_end = new_cap - suffix_len
        for i in range(suffix_len):
            new_buf[new_gap_end + i] = self.buffer[self.gap_end + i]

        self.buffer = new_buf
        self.gap_end = new_gap_end
        self.capacity = new_cap

    def move_cursor(self, pos: int) -> None:
        assert 0 <= pos <= len(self)
        if pos < self.gap_start:
            shift = self.gap_start - pos
            while shift > 0:
                self.gap_start -= 1
                self.gap_end -= 1
                self.buffer[self.gap_end] = self.buffer[self.gap_start]
                shift -= 1
        elif pos > self.gap_start:
            shift = pos - self.gap_start
            while shift > 0:
                self.buffer[self.gap_start] = self.buffer[self.gap_end]
                self.gap_start += 1
                self.gap_end += 1
                shift -= 1

    def insert(self, text: str) -> None:
        if self.gap_end - self.gap_start < len(text):
            self._grow(len(text))
        for c in text:
            self.buffer[self.gap_start] = c
            self.gap_start += 1

    def erase_back(self) -> bool:
        if self.gap_start == 0:
            return False
        self.gap_start -= 1
        return True

    def to_string(self) -> str:
        prefix = "".join(self.buffer[:self.gap_start])
        suffix = "".join(self.buffer[self.gap_end:self.capacity])
        return prefix + suffix


class Piece:
    __slots__ = ('buffer_type', 'start', 'length')

    def __init__(self, buffer_type: str, start: int, length: int):
        self.buffer_type = buffer_type  # 'ORIGINAL' or 'ADD'
        self.start = start
        self.length = length


class PieceTable:
    """Piece Table decoupling text into Original/Add buffers and piece sequence."""

    def __init__(self, initial_text: str = ""):
        self.original = initial_text
        self.add = ""
        self.pieces: List[Piece] = []
        if initial_text:
            self.pieces.append(Piece('ORIGINAL', 0, len(initial_text)))

    def __len__(self) -> int:
        return sum(p.length for p in self.pieces)

    def insert(self, pos: int, text: str) -> None:
        if not text:
            return
        assert 0 <= pos <= len(self)

        add_offset = len(self.add)
        self.add += text
        text_len = len(text)

        if not self.pieces:
            self.pieces.append(Piece('ADD', add_offset, text_len))
            return

        # Coalescing optimization for typing at the end
        if (pos == len(self) and self.pieces[-1].buffer_type == 'ADD' and
                self.pieces[-1].start + self.pieces[-1].length == add_offset):
            self.pieces[-1].length += text_len
            return

        curr_pos = 0
        for i, p in enumerate(self.pieces):
            next_pos = curr_pos + p.length

            if pos == curr_pos:
                self.pieces.insert(i, Piece('ADD', add_offset, text_len))
                return
            elif curr_pos < pos < next_pos:
                left_len = pos - curr_pos
                right_len = p.length - left_len

                left_piece = Piece(p.buffer_type, p.start, left_len)
                new_piece = Piece('ADD', add_offset, text_len)
                right_piece = Piece(p.buffer_type, p.start + left_len, right_len)

                self.pieces[i] = left_piece
                self.pieces.insert(i + 1, new_piece)
                self.pieces.insert(i + 2, right_piece)
                return
            curr_pos = next_pos

        self.pieces.append(Piece('ADD', add_offset, text_len))

    def erase(self, pos: int, length: int) -> None:
        if length == 0 or not self.pieces:
            return
        assert pos + length <= len(self)

        curr_pos = 0
        new_pieces: List[Piece] = []

        for p in self.pieces:
            p_start = curr_pos
            p_end = curr_pos + p.length

            if p_end <= pos or p_start >= pos + length:
                new_pieces.append(p)
            else:
                if pos > p_start:
                    new_pieces.append(Piece(p.buffer_type, p.start, pos - p_start))
                if pos + length < p_end:
                    skipped = (pos + length) - p_start
                    new_pieces.append(Piece(p.buffer_type, p.start + skipped, p.length - skipped))
            curr_pos = p_end

        self.pieces = new_pieces

    def to_string(self) -> str:
        res = []
        for p in self.pieces:
            buf = self.original if p.buffer_type == 'ORIGINAL' else self.add
            res.append(buf[p.start:p.start + p.length])
        return "".join(res)


class TestTextStructures(unittest.TestCase):
    def test_basic_gap_buffer(self):
        gb = GapBuffer(8)
        gb.insert("Hello")
        gb.insert(" ")
        gb.insert("World")
        self.assertEqual(gb.to_string(), "Hello World")

        gb.move_cursor(5)
        gb.insert(" Beautiful")
        self.assertEqual(gb.to_string(), "Hello Beautiful World")

        gb.move_cursor(len(gb))
        gb.erase_back()
        self.assertEqual(gb.to_string(), "Hello Beautiful Worl")

    def test_basic_piece_table(self):
        pt = PieceTable("The brown fox")
        pt.insert(4, "quick ")
        self.assertEqual(pt.to_string(), "The quick brown fox")

        pt.insert(len(pt), " jumps")
        self.assertEqual(pt.to_string(), "The quick brown fox jumps")

        pt.erase(4, 6)
        self.assertEqual(pt.to_string(), "The brown fox jumps")

    def test_differential_fuzzing(self):
        oracle = "Initial Document Content for Testing"
        pt = PieceTable(oracle)
        gb = GapBuffer(64)
        gb.insert(oracle)

        rng = random.Random(42)
        for step in range(500):
            op = rng.randint(0, 1)

            if op == 0:
                pos = 0 if not oracle else rng.randint(0, len(oracle))
                text = f" [edit_{step}] "

                oracle = oracle[:pos] + text + oracle[pos:]
                pt.insert(pos, text)

                gb.move_cursor(pos)
                gb.insert(text)
            else:
                if oracle:
                    pos = rng.randint(0, len(oracle) - 1)
                    max_len = min(len(oracle) - pos, 8)
                    length = rng.randint(1, max_len)

                    oracle = oracle[:pos] + oracle[pos + length:]
                    pt.erase(pos, length)

                    gb.move_cursor(pos + length)
                    for _ in range(length):
                        gb.erase_back()

            self.assertEqual(len(pt), len(oracle))
            self.assertEqual(len(gb), len(oracle))
            self.assertEqual(pt.to_string(), oracle)
            self.assertEqual(gb.to_string(), oracle)


if __name__ == '__main__':
    unittest.main()
